### PR TITLE
Optimize RLE decoding (part 4)

### PR DESCRIPTION
**Pull request by achille-roussel**
This PR is a follow up to https://github.com/segmentio/parquet-go/pull/230, it moves the unpacking routines from `encoding/delta` to `internal/bitpack` to reuse the functions in `encoding/rle` and optimize the decoding of dictionaries:

```
name                         old time/op    new time/op     delta
Decode/RLE/int32               25.4µs ± 1%      1.1µs ± 0%    -95.48%  (p=0.000 n=8+10)
Decode/RLE_DICTIONARY/int32    25.0µs ± 0%      1.1µs ± 0%    -95.41%  (p=0.000 n=10+10)

name                         old speed      new speed       delta
Decode/RLE/int32             1.57GB/s ± 1%  34.80GB/s ± 0%  +2110.13%  (p=0.000 n=8+10)
Decode/RLE_DICTIONARY/int32  1.60GB/s ± 0%  34.87GB/s ± 0%  +2078.80%  (p=0.000 n=10+10)

name                         old value/s    new value/s     delta
Decode/RLE/int32                 394M ± 1%      8702M ± 0%  +2110.13%  (p=0.000 n=8+10)
Decode/RLE_DICTIONARY/int32      400M ± 0%      8717M ± 0%  +2078.79%  (p=0.000 n=10+10)
```

Contributes to resolving https://github.com/segmentio/parquet-go/issues/226

### CODE REVIEW

File: internal/bitpack/unpack.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package bitpack
+
+// Padding is the padding expected to exist after the end of input buffers for
+// the unpacking algorithms to avoid reading beyond the end of the input.
+const Padding = 16
+
+// UnpackInt32 unpackes 32 bit integers from src to dst.
```
Comment (kevinburkesegment) [inline internal/bitpack/unpack.go:7]:
```suggestion
// UnpackInt32 unpacks 32 bit integers from src to dst.
```

File: internal/bitpack/unpack.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+//
+// The function unpacked len(dst) integers, it panics if src is too short to
+// contain len(dst) values of the given bit width.
+func UnpackInt64(dst []int64, src []byte, bitWidth uint) {
+	assertUnpack(src, len(dst), bitWidth)
+	unpackInt64(dst, src, bitWidth)
+}
+
+func assertUnpack(src []byte, count int, bitWidth uint) {
+	_ = src[:ByteCount(bitWidth*uint(count)+8*Padding)]
```
Comment (kevinburkesegment) [inline internal/bitpack/unpack.go:26]: I guess the idea is this panics if src is too short?

**Review by kevinburkesegment — APPROVED** [review]

File: internal/bitpack/unpack.go (hunk 2 above)
Comment (achille-roussel) [reply internal/bitpack/unpack.go:26]: yes

**Comment by achille-roussel** [comment]
```
name                old time/op    new time/op     delta
Decode/RLE/boolean     105ns ± 1%      102ns ± 1%     -3.00%  (p=0.000 n=10+10)
Decode/RLE/levels     17.8µs ± 0%      0.7µs ± 0%    -96.13%  (p=0.000 n=9+9)
Decode/RLE/int32      25.3µs ± 0%      1.2µs ± 0%    -95.23%  (p=0.000 n=9+9)

name                old speed      new speed       delta
Decode/RLE/boolean  11.9GB/s ± 1%   12.2GB/s ± 1%     +3.11%  (p=0.000 n=10+10)
Decode/RLE/levels    563MB/s ± 0%  14535MB/s ± 0%  +2481.21%  (p=0.000 n=9+9)
Decode/RLE/int32    1.58GB/s ± 0%  33.11GB/s ± 1%  +1997.38%  (p=0.000 n=9+9)

name                old value/s    new value/s     delta
Decode/RLE/boolean     94.8G ± 1%      97.8G ± 1%     +3.11%  (p=0.000 n=10+10)
Decode/RLE/levels       563M ± 0%     14537M ± 0%  +2481.21%  (p=0.000 n=9+9)
Decode/RLE/int32        395M ± 0%      8278M ± 1%  +1997.38%  (p=0.000 n=9+9)
```

**Pull request merged**